design/object-bucket/object-bucket.md

```diff
@@ -0,0 +1,127 @@
+# Rook-Ceph Object Bucket Controller
+
+## Proposed Feature
+
+Currently, the automation exists in the Rook-Ceph operator to support [Ceph Object Store](../object-store.md) and [Ceph Object Store User](https://github.com/rook/rook/issues/1583) custom resource definitions (CRDs).  
+
+This design closes the loop for Rook-Ceph object stores with a CephObjectBucket controller and CRD.  Users will request buckets via the Kubernetes API and have returned ConfigMaps with connection information for that bucket.  One ConfigMap will be created per CephObjectBucket.  ConfigMap names will be deterministic and based on the name of the CephObjectBucket.  This gives users the opportunity to define their Pod spec at the same time as the CephObjectBucket.  Additionally, the CephObjectBucket and Pod can safely be deployed at the same time thanks to built in synchronization in Kubernetes.
+
+Bucket deletion in the early stages of this controller design will not be addressed beyond cleaning up generated resources.  Deleting a CephObjectBucket will only cause deletion of the associated ConfigMap.  The controller will not delete the bucket from the object store. 
+
+## Dependencies
+
+This design assumes that cross-namespace support for CephObjectUsers, proposed in #2473, has been accepted into the project.  Specifically, the coupling of CephObjectStores to StorageClasses, which are then referenced by the CephObjectUser.
+
+## Object Bucket Walkthrough
+
+### Assumptions
+
+- The `storageClass` field has been added to the CephObjectStoreUser API
```

Comment:
is this strictly necessary to implement this new bucket CRD and controller? it looks like this proposal could still be implemented without the `StorageClass` being added to the `CephObjectStoreUser` CRD.

Reply:
@jbw976 Addind storage class to the user CRD is not necessary for the bucket design. The bucket CRD just needs to know the object store service and secret. As long as the name of the key-pair secret is deterministic the `CephObjectStoreUser` is not required at all.